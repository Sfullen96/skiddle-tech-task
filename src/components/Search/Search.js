import React, { useRef } from 'react';
import PropTypes from 'prop-types';
import { AsyncTypeahead } from 'react-bootstrap-typeahead';
import { Form, Button } from 'react-bootstrap';
import moment from 'moment';

import './Search.scss';
import logo from '../../images/skiddle-logo-white-stacked.png';

/**
 * Search
 *
 * The component that contains the search box. Uses module react-bootstrap-typeahead
 * (https://github.com/ericgio/react-bootstrap-typeahead) to provide async searching of
 * the API, as well as debouncing.
 */
const Search = ({
  doSearch,
  isFetching,
  searchResults,
  handleDropdownOptionSelection,
  onSubmit,
  onKeyDown,
}) => {
  // Generate a ref to the typeahead comp so we can explicitly blur
  // when the enter key is pressed - see 'onKeyDown' func in SearchContainer.js
  const typeahead = useRef(null);

  /**
   * @function generateMenuItem
   *
   * Generate a custom dropdown list item
   * @param {Object} option
   */
  const generateMenuItem = option => (
    <div className="search__input-container--menu-item">
      <p className="mb-0">
        <strong>{option.eventname}</strong>
      </p>
      <p>
        <small>
          {moment(option.date).format('DD/MM/YYYY')} @ {option.venue.name}, {option.venue.town}
        </small>
      </p>
    </div>
  );

  return (
    <div className="search__search-bar-container mb-4">
      <div className="search__overlay" />
      <div className="search__input-container">
        <img src={logo} alt="Skiddle Logo" className="search__input-container--logo" />
        <Form onSubmit={onSubmit}>
          <AsyncTypeahead
            onSearch={e => doSearch(e)}
            className="search__input-container--input"
            isLoading={isFetching}
            options={searchResults}
            // Must have typed 3 characters before attempting search
            // to avoid too many results
            minLength={3}
            onChange={option => handleDropdownOptionSelection(option.id)}
            placeholder="Search for Events..."
            id="search"
            delay={500} // How long after typing has stopped before searching
            labelKey={option =>
              `${option.eventname} - ${moment(option.date).format('DD/MM/YYYY')} @${
                option.venue.name
              }, ${option.venue.town}`
            }
            maxResults={5}
            renderMenuItemChildren={option => generateMenuItem(option)}
            onKeyDown={e => onKeyDown(e, typeahead)}
            ref={typeahead}
          />
          <Button variant="outline-light" type="submit" className="search__input-container--submit">
            <i className="fa fa-search" />
          </Button>
        </Form>
      </div>
    </div>
  );
};

Search.defaultProps = {
  searchResults: [],
};

Search.propTypes = {
  doSearch: PropTypes.func.isRequired,
  isFetching: PropTypes.bool.isRequired,
  searchResults: PropTypes.arrayOf(
    PropTypes.shape({
      error: PropTypes.bool,
      pageCount: PropTypes.number,
      results: PropTypes.arrayOf(
        PropTypes.shape({
          id: PropTypes.number,
          date: PropTypes.string,
          venue: PropTypes.shape({
            name: PropTypes.string,
            town: PropTypes.string,
          }),
          eventname: PropTypes.string,
        }),
      ),
      totalCount: PropTypes.number,
    }),
  ),
  handleDropdownOptionSelection: PropTypes.func.isRequired,
  onSubmit: PropTypes.func.isRequired,
  onKeyDown: PropTypes.func.isRequired,
};

export default Search;
