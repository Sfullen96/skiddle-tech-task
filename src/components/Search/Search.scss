@import '../../_globals.scss';

.search {
  &__search-bar-container {
    position: relative;
    background-image: url(../../images/music-crowd-concert-stock-2018-u-billboard-1548.jpg);
    background-size: cover;
    background-position-y: -300px;
    height: 0;
    width: 100%;
    background-repeat: no-repeat;
    min-height: 500px;
    width: 100%;
    float: left;
    border: 1px solid $primary-border;
  }
  &__overlay {
    position: absolute;
    width: 100%;
    height: 100%;
    background-color: rgba(0, 0, 0, 0.5);
  }
  &__input-container {
    position: absolute;
    width: 50%;
    top: 40%;
    left: 50%;
    transform: translate(-50%, -50%);
    text-align: center;
    &--logo {
      width: 300px;
    }
    &--input {
      width: 90%;
      float: left;
      input {
        border-radius: 0;
      }
    }
    &--submit {
      width: 10%;
      float: left;
      border-radius: 0;
    }
    &--menu-item {
      border-bottom: 1px solid rgba(0, 0, 0, 0.1);
    }
  }
  &__results {
    &--result {
      text-align: center;
      .figure {
        margin-bottom: 20px;
        border: 1px solid $primary-border;
        background-color: #fff;
        padding: 10px 15px;
        // min-height: 425px;
        .figure-image-container {
          width: 100%;
          float: left;
          height: 250px;
          .figure-img {
            border: 3px solid $primary-border;
            margin-bottom: 0;
            width: auto;
            height: 100%;
          }
        }
        .figure-caption-container {
          padding-left: 15px;
          padding-right: 15px;
          padding-top: 15px;
          width: 100%;
          float: left;
          height: 100px;
          .figure-caption {
            height: 150px;
            float: left;
            width: 100%;
            p {
              width: 100%;
              float: left;
            }
            .event-name {
              font-size: 15px;
              font-weight: 800;
            }
          }
        }
        .event-more-info-container {
          width: 100%;
          float: left;
          height: 80px;
          button {
            border-radius: 0;
          }
        }
      }
    }
  }
}
