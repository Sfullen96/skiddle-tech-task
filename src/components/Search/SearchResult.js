import React from 'react';
import PropTypes from 'prop-types';
import { Figure, Button } from 'react-bootstrap';
import moment from 'moment';
import { Link } from 'react-router-dom';

/**
 * SearchResult
 *
 * The individual search result tile
 */
const SearchResult = ({ image, venue, id, name, date }) => (
  <Figure>
    <div className="figure-image-container">
      <Figure.Image width="100%" height="220" src={image} thumbnail />
    </div>
    <div className="figure-caption-container">
      <Figure.Caption>
        <p className="event-name">{name}</p>
        <p className="event-info">
          {moment(date).format('DD/MM/YYYY')} @ {venue.name}, {venue.town}
        </p>
      </Figure.Caption>
    </div>
    <div className="event-more-info-container">
      <hr />
      <Link to={`/event/${id}`}>
        <Button variant="outline-dark" type="button">
          More Info <i className="fa fa-chevron-right" />
        </Button>
      </Link>
    </div>
  </Figure>
);

SearchResult.propTypes = {
  image: PropTypes.string.isRequired,
  venue: PropTypes.shape({
    name: PropTypes.string,
    town: PropTypes.string,
  }).isRequired,
  id: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  date: PropTypes.string.isRequired,
};

export default SearchResult;
