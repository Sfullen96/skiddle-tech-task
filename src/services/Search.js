import camelCaseKeys from 'camelcase-keys-recursive';

import { Endpoints } from '../constants';
import { getRequest } from '../helpers';

/**
 * @function search
 *
 * The function which calls the Skiddle API to fetch event data based
 * on a keyword and offset
 *
 * @param {string} keyword - The search term
 * @param {number} offset - Used to paginate results
 */
const search = async (keyword, offset = 0) => {
  try {
    const { data } = await getRequest(Endpoints.SEARCH, {}, { keyword, offset });
    const { error, pagecount, results, totalcount } = data;
    return camelCaseKeys({
      error,
      pageCount: pagecount,
      results,
      totalCount: parseInt(totalcount, 10),
    });
  } catch (e) {
    return e;
  }
};

export default search;
