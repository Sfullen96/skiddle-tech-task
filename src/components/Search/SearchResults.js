import React from 'react';
import PropTypes from 'prop-types';
import ReactPaginate from 'react-paginate';
import { Container, Row, Col } from 'react-bootstrap';

import { Loader } from '..';

import SearchResult from './SearchResult';

/**
 * SearchResults
 *
 * The component that displays the main search results, hidden until a user
 * submits the form via the search button or pressing enter when focused on the
 * input
 */
const SearchResults = ({
  searchResults,
  searchTerm,
  totalPages,
  totalResults,
  changePage,
  isFetching,
}) => (
  <Container className="search__results">
    {isFetching && <Loader />}
    {!searchResults.length ? (
      <h2>There are no results for &quot;{searchTerm}&quot;</h2>
    ) : (
      <>
        <h2 className="mb-4">
          Found {totalResults} results for &quot;{searchTerm}&quot;
        </h2>
        <Row>
          {searchResults.map(result => (
            <Col xs="12" sm="6" md="4" className="search__results--result" key={result.id}>
              <SearchResult
                image={result.largeimageurl}
                venue={result.venue}
                id={result.id}
                name={result.eventname}
                date={result.date}
              />
            </Col>
          ))}
        </Row>
        <Row>
          <ReactPaginate
            previousLabel="previous"
            nextLabel="next"
            breakLabel="..."
            breakClassName="break"
            pageCount={totalPages}
            marginPagesDisplayed={2}
            pageRangeDisplayed={5}
            onPageChange={changePage}
            containerClassName="pagination"
            subContainerClassName="pages pagination"
            activeClassName="active"
            pageClassName="page-item"
            pageLinkClassName="page-link"
            previousClassName="page-item"
            nextClassName="page-item"
            previousLinkClassName="page-link"
            nextLinkClassName="page-link"
          />
        </Row>
      </>
    )}
  </Container>
);

SearchResults.defaultProps = {
  searchResults: [],
  searchTerm: null,
  totalPages: null,
  totalResults: null,
  isFetching: false,
};

SearchResults.propTypes = {
  searchResults: PropTypes.arrayOf(
    PropTypes.shape({
      error: PropTypes.bool,
      pageCount: PropTypes.number,
      results: PropTypes.arrayOf(
        PropTypes.shape({
          id: PropTypes.number,
          largeimageurl: PropTypes.string,
          date: PropTypes.string,
          venue: PropTypes.shape({
            name: PropTypes.string,
            town: PropTypes.string,
          }),
          eventname: PropTypes.string,
        }),
      ),
      totalCount: PropTypes.number,
    }),
  ),
  searchTerm: PropTypes.string,
  totalPages: PropTypes.number,
  totalResults: PropTypes.number,
  changePage: PropTypes.func.isRequired,
  isFetching: PropTypes.bool,
};

export default SearchResults;
