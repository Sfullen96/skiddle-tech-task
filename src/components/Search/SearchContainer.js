import React, { useState } from 'react';
import { Redirect } from 'react-router-dom';

import './Search.scss';

import { search } from '../../services';
import Search from './Search';
import SearchResults from './SearchResults';

/**
 * SearchContainer
 *
 * A component which encompasses the search field and the search results
 * components, and controls all the stateful logic
 */
const SearchContainer = () => {
  // Declare our initial state variables using the 'useState' hook
  const [isFetching, setIsFetching] = useState(false);
  const [searchResults, setSearchResults] = useState([]);
  const [searchTerm, setSearchTerm] = useState(null);
  const [totalPages, setTotalPages] = useState(null);
  const [totalResults, setTotalResults] = useState(null);
  const [redirect, setRedirect] = useState(null);
  const [showResults, setShowResults] = useState(false);
  const resultsPerPage = 20;

  // If state contains a redirect, send them to it
  // the 'push' prop pushes to history to allow backwards
  // navigation in browser
  if (redirect) {
    return <Redirect to={redirect} push />;
  }

  /**
   * @function doSearch
   *
   * The actual search function, handles the search and the setting
   * of results in state.
   *
   * @param {String} keyword - The user-typed search term itself
   * @param {Number} nextPage - The next page, used to calculate offset
   */
  const doSearch = async (keyword, nextPage) => {
    // Set the loading state
    setIsFetching(true);
    // Keep the search term in state to be used in displaying e.g.
    // Found 10 results for "{searchTerm}"
    setSearchTerm(keyword);

    // Calculate the offset, defaulting to 0
    let offset = 0;
    if (nextPage) {
      offset = (nextPage - 1) * resultsPerPage;
    }

    // Actually fetch the data
    const { error, results, totalCount } = await search(keyword, offset);

    // ToDo: Improve error handling with a message
    if (error) {
      setIsFetching(false);
      setSearchResults([]);
      return;
    }

    // Set the results in state
    setTotalPages(Math.ceil(totalCount / resultsPerPage));
    setSearchResults(results);
    setTotalResults(totalCount);
    setIsFetching(false);
  };

  /**
   * @function handleDropdownOptionSelection
   *
   * Handle a selection being chosen from the dropdown to
   * redirect them to the event they clicked
   * @param {Number} eventId - The event ID
   */
  const handleDropdownOptionSelection = eventId => {
    setRedirect(`/event/${eventId}`);
  };

  /**
   * @function onSubmit
   *
   * Triggered when a user presses the maginfying glass button, at this point
   * there are already search results in state so we simply need to allow them
   * to be shown by setting a variable in state
   * @param {*} e - The event
   */
  const onSubmit = e => {
    e.preventDefault();
    setShowResults(true);
  };

  /**
   * @function changePage
   *
   * Handle changing page, either backwards or forwards
   * @param {*} current - The current page
   */
  const changePage = async current => doSearch(searchTerm, parseInt(current.selected + 1, 10));

  /**
   * @function onKeyDown
   *
   * Check if the user presses the enter key (code 13). <AsyncTypeahead /> provides no way
   * to submit the form on enter, and as we want the form to be submitted and the main results
   * shown under the search field on enter press we need this workaround
   * @param {*} e - The event
   * @param {*} typeahead - The ref to the typeahead component
   */
  const onKeyDown = async (e, typeahead) => {
    if (e.keyCode === 13) {
      setSearchTerm(e.target.value);
      doSearch(e.target.value);
      // Manually blur the input
      typeahead.current.getInstance().blur();
      // Show the main results
      setShowResults(true);
    }
  };

  return (
    <div className="search">
      <Search
        doSearch={doSearch}
        isFetching={isFetching}
        searchResults={searchResults}
        searchTerm={searchTerm}
        totalPages={totalPages}
        totalResults={totalResults}
        showResults={showResults}
        resultsPerPage={20}
        handleDropdownOptionSelection={handleDropdownOptionSelection}
        onSubmit={onSubmit}
        onKeyDown={onKeyDown}
      />
      {showResults && (
        <SearchResults
          searchResults={searchResults}
          searchTerm={searchTerm}
          totalPages={totalPages}
          totalResults={totalResults}
          changePage={changePage}
          isFetching={isFetching}
        />
      )}
    </div>
  );
};

export default SearchContainer;
