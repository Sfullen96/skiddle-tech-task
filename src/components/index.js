import Header from './Header/Header';
import SearchContainer from './Search/SearchContainer';
import Loader from './Loader/Loader';
import EventDetails from './EventDetails/EventDetails';

export { Header, SearchContainer, Loader, EventDetails };
