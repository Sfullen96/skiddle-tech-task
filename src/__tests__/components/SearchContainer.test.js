import React from 'react';
import { shallow, mount } from 'enzyme';

import { SearchContainer } from '../../components';
import Search from '../../components/Search/Search';
import SearchResults from '../../components/Search/SearchResults';

describe('<SearchContainer />', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallow(<SearchContainer />);
  });

  it('should match the snapshot', () => expect(wrapper).toMatchSnapshot());

  it('should render without crashing', () => {
    expect(wrapper.find('div.search').length).toEqual(1);
  });

  it('should contain the <Search /> component', () => {
    expect(wrapper.find('Search').length).toEqual(1);
  });
});

describe('Mounted <SearchContainer />', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = mount(<SearchContainer />);
  });

  it('should contain the <SearchResults /> component once a search is submitted', () => {
    expect(wrapper.find('SearchResults').length).toEqual(0);
    wrapper.find('input.rbt-input').simulate('change', { target: { value: 'Test' } });
    wrapper.find('form').simulate('submit');
    expect(wrapper.find('SearchResults').length).toEqual(1);
  });
});
