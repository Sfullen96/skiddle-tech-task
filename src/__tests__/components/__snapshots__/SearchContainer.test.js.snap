// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchContainer /> should match the snapshot 1`] = `ShallowWrapper {}`;
